dpctl/tensor/libtensor/include/kernels/boolean_reductions.hpp

```diff
@@ -0,0 +1,616 @@
+          inp_reduced_dims_indexer_(arg_reduced_dims_indexer),
+          reduction_max_gid_(reduction_size)
+    {
+    }
+
+    void operator()(sycl::id<1> id) const
+    {
+
+        auto inp_out_iter_offsets_ = inp_out_iter_indexer_(id[0]);
+        const auto &inp_iter_offset = inp_out_iter_offsets_.get_first_offset();
+        const auto &out_iter_offset = inp_out_iter_offsets_.get_second_offset();
+
+        outT red_val(identity_);
+        for (size_t m = 0; m < reduction_max_gid_; ++m) {
+            auto inp_reduction_offset = inp_reduced_dims_indexer_(m);
+            auto inp_offset = inp_iter_offset + inp_reduction_offset;
+
+            // must convert to boolean first to handle nans
+            using dpctl::tensor::type_utils::convert_impl;
+            outT val = convert_impl<bool, argT>(inp_[inp_offset]);
+
+            red_val = reduction_op_(red_val, val);
+        }
+
+        out_[out_iter_offset] = red_val;
+    }
+};
+
+template <typename argT, typename outT, typename ReductionOp, typename GroupOp>
+struct ContigBooleanReduction
+{
+private:
+    const argT *inp_ = nullptr;
+    outT *out_ = nullptr;
+    ReductionOp reduction_op_;
+    GroupOp group_op_;
+    size_t reduction_max_gid_ = 0;
+    size_t reductions_per_wi = 16;
+
+public:
+    ContigBooleanReduction(const argT *inp,
+                           outT *res,
+                           ReductionOp reduction_op,
+                           GroupOp group_op,
+                           size_t reduction_size,
+                           size_t reduction_size_per_wi)
+        : inp_(inp), out_(res), reduction_op_(reduction_op),
+          group_op_(group_op), reduction_max_gid_(reduction_size),
+          reductions_per_wi(reduction_size_per_wi)
+    {
+    }
+
+    void operator()(sycl::nd_item<2> it) const
+    {
+
+        size_t reduction_id = it.get_group(0);
+        size_t reduction_batch_id = it.get_group(1);
+
+        auto work_group = it.get_group();
+        size_t wg_size = it.get_local_range(1);
+
+        size_t base = reduction_id * reduction_max_gid_;
+        size_t start = base + reduction_batch_id * wg_size * reductions_per_wi;
+        size_t end = std::min((start + (reductions_per_wi * wg_size)),
+                              base + reduction_max_gid_);
+
+        // reduction to the work group level is performed
+        // inside group_op
+        outT red_val_over_wg = group_op_(work_group, inp_ + start, inp_ + end);
+
+        if (work_group.leader()) {
+            sycl::atomic_ref<outT, sycl::memory_order::relaxed,
+                             sycl::memory_scope::device,
+                             sycl::access::address_space::global_space>
+                res_ref(out_[reduction_id]);
+            outT read_val = res_ref.load();
+            outT new_val{};
+            do {
+                new_val = reduction_op_(read_val, red_val_over_wg);
+            } while (!res_ref.compare_exchange_strong(read_val, new_val));
```

Comment:
I think it is going to be few times faster to use `res_ref.fetch_and(red_val_over_wg)` or `res_ref.fetch_or(red_val_over_wg)` depending on the operation.